server/storage/players.lua

```diff
@@ -382,8 +418,99 @@ local function cleanPlayerGroups()
+        return
+    end
+
+    if not jsonPlayerFields[key] and subKeys then
+        error(('Tried to update player data field %s as a json object when it isn\'t one'):format(key))
+        return
+    end
+
+    value = type(value) == 'table' and json.encode(value) or value
+
+    -- In sendPlayerDataUpdates we don't go more than 3 tables deep
+    if subKeys and #subKeys > 3 then
+        error(('Cannot save field %s because data is too big.\nsubKeys: %s\nvalue: %s'):format(key, json.encode(subKeys), value))
+        return
+    end
+
+    local currentTable = isUpdating and playerDataUpdateQueue or collectedPlayerData
+
+    if not currentTable[citizenid] then
+        currentTable[citizenid] = {}
+    end
+
+    currentTable[citizenid][key] = subKeys and {} or value
+
+    if subKeys then
+        local current = currentTable[citizenid][key]
+        -- We don't check the last one because otherwise we lose the table reference
+        for i = 1, #subKeys - 1 do
+            if not current[subKeys[i]] then
+                current[subKeys[i]] = {}
+            end
+        end
+
+        current[subKeys[#subKeys]] = value
+    end
+end
+
+local function sendPlayerDataUpdates()
+    -- We implement this to ensure when updating no values are added to our updating sequence to prevent data loss by accidentally skipping over it
+    isUpdating = true
+
+    for citizenid, playerData in pairs(collectedPlayerData) do
+        for key, data in pairs(playerData) do
+            if type(data) == 'table' then
+                local updateStrings = {}
+                -- We go a maximum of 3 tables deep into the current table to prevent misuse and qbox doesn't have more than 2 actually
+                -- If we were to make this variable to the amount of data there is, then enough tables can crash the server
+                for k, v in pairs(data) do
+                    if type(v) == 'table' then
+                        for k2, v2 in pairs(v) do
+                            if type(v2) == 'table' then
+                                for k3, v3 in pairs(v2) do
+                                    if type(v3) == 'table' then
+                                        v3 = json.encode(v3)
+                                    end
+
+                                    updateStrings[#updateStrings + 1] = { ('$.%s.%s.%s'):format(k, k2, k3), v3 }
+                                end
+                            else
+                                updateStrings = { ('$.%s.%s'):format(k, k2), v2 }
+                            end
+                        end
+                    else
+                        updateStrings[#updateStrings + 1] = { ('$.%s'):format(k), v }
+                    end
+                end
+
+                for i = 1, #updateStrings do
+                    MySQL.prepare.await(('UPDATE players SET %s = JSON_SET(%s, "%s", ?) WHERE citizenid = ?'):format(key, key, updateStrings[i][1]), { updateStrings[i][2], citizenid })
```

Comment:
Is it better to loop and make multiple requests like this, or maybe to create a bulk request will all the changes at once? Or maybe some sort of combination of both?

Reply:
The idea of a transaction slipped my mind. Prepared statements just seemed the best since it will update frequently, but it's not very performant against a transaction possibly.